test/dist_autograd_test.py

```diff
@@ -1253,5 +1253,68 @@ def test_backwards_nested_python_udf(self):
             self.assertEqual(t1.grad, grads[t1])
             self.assertEqual(t2.grad, grads[t2])
 
+    _my_backward_func_executed = False
+
+    class MyBackwardFunc(Function):
+        @staticmethod
+        def forward(ctx, input):
+            return input
+
+        @staticmethod
+        @once_differentiable
+        def backward(ctx, input):
+            DistAutogradTest._my_backward_func_executed = True
+            return input
+
+    def _clear_context_thread(context):
+        # Wait until MyBackwardFunc is executed and then clean up context. This
+        # ensures we simulate a case where we clean up the context while the
+        # backward pass is running.
+        while not DistAutogradTest._my_backward_func_executed:
+            time.sleep(0.1)
+        dist_autograd._release_context(context._context_id())
+
+    @dist_init
+    def test_clean_context_during_backward(self):
+        '''
+        This test simulates the situation where the 'backward' call might throw
+        an exception locally which would lead to the autograd context being
+        cleaned up if we're using the context manager. As a result, the autograd
+        context might be cleaned up while some threads are still using the
+        autograd context.
+
+        It is fine for the 'backward' call to throw an exception in this test,
+        but the process should not crash.
+        '''
+
+        context = dist_autograd._new_context()
+        t1 = torch.rand((3, 3), requires_grad=True)
+        for i in range(0, 100):
+            if i == 50:
+                # Call MyBackwardFunc somewhere in the middle of the autograd
+                # graph.
+                t1 = DistAutogradTest.MyBackwardFunc.apply(t1)
+            t1 = rpc.rpc_sync("worker{}".format(self._next_rank()), torch.add, args=(t1, t1))
+        self.assertEqual(100, len(context._send_functions()))
+
+        # Start thread to cleanup context.
+        t = threading.Thread(target=DistAutogradTest._clear_context_thread, args=(context,))
+        t.start()
+
+        with self.assertRaisesRegex(RuntimeError, "Could not find autograd context with id"):
+            dist_autograd.backward([t1.sum()])
```

Comment:
I'm not 100% clear on what behavior is being tested here and what should be different than from before this PR. Does this result in the backward call throwing an exception (if so, what is exactly is the exception that should be thrown)?

Reply:
The exception being thrown is "Could not find autograd context id for". This happens because  the autograd context is cleaned up and some thread on some node is still looking for that autograd context.

Before this PR, this test would cause the process to crash since a thread would be using a reference to `DistAutogradContext` but our clean up logic would delete the context from the container.

Reply:
Does this assume that the thread cleared the context before the backward uses it? How do we guarantee that?